File: mono/bot.cs
using System;
using System.IO;
using System.Net.Sockets;
using Newtonsoft.Json;

public class Bot {
	public static void Main(string[] args) {
	    string host = args[0];
        int port = int.Parse(args[1]);
        string botName = args[2];
        string botKey = args[3];

		Console.WriteLine("Connecting to " + host + ":" + port + " as " + botName + "/" + botKey);

		using(TcpClient client = new TcpClient(host, port)) {
			NetworkStream stream = client.GetStream();
			StreamReader reader = new StreamReader(stream);
			StreamWriter writer = new StreamWriter(stream);
			writer.AutoFlush = true;

			new Bot(reader, writer, new Join(botName, botKey));
		}
	}

	private StreamWriter writer;

	Bot(StreamReader reader, StreamWriter writer, Join join) {
		this.writer = writer;
		string line;

		send(join);

		while((line = reader.ReadLine()) != null) {
			MsgWrapper msg = JsonConvert.DeserializeObject<MsgWrapper>(line);
			switch(msg.msgType) {
				case "carPositions":
					send(new Throttle(0.5));
					break;
				case "join":
					Console.WriteLine("Joined");
					send(new Ping());
					break;
				case "gameInit":
					Console.WriteLine("Race init");
					send(new Ping());
					break;
				case "gameEnd":
					Console.WriteLine("Race ended");
					send(new Ping());
					break;
				case "gameStart":
					Console.WriteLine("Race starts");
					send(new Ping());
					break;
				default:
					send(new Ping());
					break;
			}
		}
	}

	private void send(SendMsg msg) {
		writer.WriteLine(msg.ToJson());
	}
}

class MsgWrapper {
    public string msgType;
    public Object data;

    public MsgWrapper(string msgType, Object data) {
    	this.msgType = msgType;
    	this.data = data;
    }
}

abstract class SendMsg {
	public string ToJson() {
		return JsonConvert.SerializeObject(new MsgWrapper(this.MsgType(), this.MsgData()));
	}
	protected virtual Object MsgData() {
        return this;
    }

    protected abstract string MsgType();
}

class Join: SendMsg {
	public string name;
	public string key;
	public string color;

	public Join(string name, string key) {
		this.name = name;
		this.key = key;
		this.color = "red";
	}

	protected override string MsgType() { 
		return "join";
	}
}

class Ping: SendMsg {
	protected override string MsgType() {
		return "ping";
	}
}

class Throttle: SendMsg {
	public double value;

	public Throttle(double value) {
		this.value = value;
	}

	protected override Object MsgData() {
		return this.value;
	}

	protected override string MsgType() {
		return "throttle";
	}
}

File: fsharp/bot.fs
open System
open System.IO
open System.Net.Sockets
open FSharp.Data

type Json = JsonProvider<"./sample.json", SampleIsList=true, RootName="message">

let send (msg : Json.Message) (writer : StreamWriter) =
    writer.WriteLine (msg.JsonValue.ToString(JsonSaveOptions.DisableFormatting))
    writer.Flush()

let join name key color =
    Json.Message(msgType = "join", data = Json.DecimalOrData(Json.Data(name = Some(name), key = Some(key), color = Some(color))))

let throttle (value : Decimal) =
    Json.Message(msgType = "throttle", data = Json.DecimalOrData(value))

let ping =
    Json.Message(msgType = "ping", data = null)

[<EntryPoint>]
let main args =
    let (host, portString, botName, botKey) = 
      args |> (function [|a;b;c;d|] -> a, b, c, d | _ -> failwith "Invalid param array")
    let port = Convert.ToInt32(portString)

    printfn "Connecting to %s:%d as %s/%s" host port botName botKey

    use client = new TcpClient(host, port)
    let stream = client.GetStream()
    let reader = new StreamReader(stream)
    let writer = new StreamWriter(stream)

    send (join botName botKey "blue") writer

    let processNextLine nextLine =
        match nextLine with
        | null -> false
        | line ->
            try 
                let msg = Json.Parse(line)
                match msg.MsgType with
                | "carPositions" -> send (throttle 0.5m) writer
                | "join" -> printfn "Joined"; send (ping) writer
                | "gameInit" -> printfn "Race init"; send (ping) writer
                | "gameEnd" -> printfn "Race ends"; send (ping) writer
                | "gameStart" -> printfn "Race starts"; send (ping) writer
                | _ -> send (ping) writer
                true
            with
            | ex -> true

    while processNextLine (reader.ReadLine()) do 
        ()

    0